PlayMe.Server/AutoPlay/MultiAutoPlay/MultiAutoPlay.cs

```diff
@@ -0,0 +1,157 @@
+﻿using System;
+using System.Collections.Generic;
+using PlayMe.Common.Model;
+using PlayMe.Server.AutoPlay.MultiAutoPlay.Config;
+using PlayMe.Common.Util;
+using PlayMe.Server.AutoPlay.MultiAutoPlay;
+using System.Threading;
+using PlayMe.Plumbing.Diagnostics;
+using PlayMe.Server.Providers.NewSpotifyProvider;
+using PlayMe.Server.Queue;
+using System.Linq;
+using PlayMe.Server.AutoPlay.Util;
+
+namespace PlayMe.Server.AutoPlay.MultiAutoplay
+{
+    public class MultiAutoPlay : IAutoPlay
+    {
+        private readonly Stack<QueuedTrack> _tracksForAutoplaying = new Stack<QueuedTrack>();
+
+        private readonly IList<IWeightedAutoPlay> autoPlayRepository;
+        private readonly AutoPlayResolver autoPlayResolver;
+        private readonly NewSpotifyProvider _spotify;
+        private readonly IForbiddenMusicService _forbiddenMusicService;
+        private readonly ILogger _logger;
+
+        // Set to a reasonable number to handle veto-battles
+        private const int TRACK_CACHE_SIZE = 5;
+        private const int GET_TRACK_RETRY_LIMIT = 15;
+
+        private static Object _fillCacheLock = new Object();
+        private static Object _querySongsLock = new Object();
+
+        public MultiAutoPlay(
+            IWeightedAutoPlayRepository autoPlayRepository,
+            AutoPlayResolver autoPlayResolver,
+            NewSpotifyProvider spotify,
+            IForbiddenMusicService forbiddenMusicService,
+            ILogger logger)
+        {
+            _logger = logger;
+            _spotify = spotify;
+            _forbiddenMusicService = forbiddenMusicService;
+
+            this.autoPlayRepository = autoPlayRepository.GetAllAutoPlays();
+
+            if (this.autoPlayRepository.Count == 0)
+            {
+                throw new Exception("No autoPlay instances loaded by WeightedAutoPlayProvider.");
+            }
+
+            this.autoPlayResolver = autoPlayResolver;
+        }
+
+        public QueuedTrack FindTrack()
```

Comment:
Concurrent stack added.

The locks were initially to stop errors with the API being called synchronously across different threads.